api/v1alpha1/clienttrafficpolicy_types.go

```diff
@@ -78,6 +78,10 @@ type ClientTrafficPolicySpec struct {
 	//
 	// +optional
 	Path *PathSettings `json:"path,omitempty"`
+	// The compression config for the downstream client connections.
```

Comment:
lets define this config in one place for now, I vote to rm this one and keep the one in `backendTrafficPolicy`

Reply:
Emm... then how we distinguish the upstream and downstream side compression in the future?

Reply:
afaik compression is defined at a request/response level per listener/route in envoy https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/compressor_filter

Reply:
Yes, for the only downstream side compression, we will enable the response direction compression by `response_direction_config` and disable the request direction compression by the `request_direction_config`

For the only upstream side compression, we will enable the requestion direction compression by `request_direction_config` and diable the response direction compression by `response_direction_config`

Reply:
> Yes, for the only downstream side compression, we will enable the response direction compression by `response_direction_config` and disable the request direction compression by the `request_direction_config`

This supposed to be defined by backend traffic policy.

> 
> For the only upstream side compression, we will enable the requestion direction compression by `request_direction_config` and diable the response direction compression by `response_direction_config`

This supposed to be defined by client traffic policy. That confuses me if define this in the backend traffic policy. What we do when want to enable the dowstream side compression?

Reply:
its defined in terms of `request` and `response` and `request` is disabled by default, in the future we can enhance config and enable request compression if needed
```
compression:
  type: Gzip
  gzip:
    request:
    response:
       ......
```

Reply:
a downstream request is also a upstream request, so it cannot really be differentiated b/w ClientTraffic and BackendTraffic imo. I'd put it in `BackendTraffic` for two reasons
1. feature is relevant to app teams
2. feature can be enabled/disabled at the route level

Reply:
ah, I begin to get your point now. thanks for explanation this for me!

> a downstream request is also a upstream request, so it cannot really be differentiated b/w ClientTraffic and BackendTraffic imo. I'd put it in `BackendTraffic` for two reasons
> 
> 1. feature is relevant to app teams

May I ask how to define this? I can understand that for the upstream side, the only the application knows that whether the application has the ability to compressor and decompressor.

For the downstream side, why this is defined by the app teams? since the app teams also know the ability of the applciation?

> 2. feature can be enabled/disabled at the route level

Reply:
> since the app teams also know the ability of the applciation

yeah since they have a better understanding of their L7 client, this is just my initial thought on where to put the config, would be good to hear counter arguments as well

Reply:
> > 2. feature can be enabled/disabled at the route level
> >    +1

Will we have something like PreRouteTrafficPolicy in the future? I didn't something like this today.

Reply:
no, the `BackendTrafficPolicy` can be attached either to the `Gateway` or to the `Route`

Reply:
oop, I missed that, thanks!

Reply:
> > since the app teams also know the ability of the applciation
> 
> yeah since they have a better understanding of their L7 client, this is just my initial thought on where to put the config, would be good to hear counter arguments as well

@arkodg I got your reason for putting this in the `BackendTrafficPolicy`, but I also feel it is against the intuition of the users.

The most two common use cases for `Compression` and `Decompression` are:
* Compression: compress the response from the backend to the client to save the network bandwidth.
* Decompression: decompress the request from the client to the backend to avoid the backend to do the decompression.

From the user's point of view, these two use cases are more related to client-proxy communication. I guess the real issue here is that the ClientTrafficPolicy is not able to do per-route configuration.

Reply:
imo the nouns we could have come up with were
* trafficPolicy - to deal with L7 request/response transformations
* clientConnectionPolicy - for client connection
* backendConnectionPolicy - for backend connection

`backendTrafficPolicy` imo is combining `backendConnectionPolicy` + `trafficPolicy`

Reply:
> imo the nouns we could have come up with were
> 
> * trafficPolicy - to deal with L7 request/response transformations

Thinking of the role for the compression API. I feel there are two rules may involve this.

One is the application developer, just as we discussed above, they know the capabilities of the application.
The second one is the cluster operator, like the case of @zhaohuabing said, the cluster operator wants to save the bandwidth.

But for the second case, it should be under the service mesh case. There is sidecar proxy in the front of the backend endpoint. Then cluster operator needn't care about the application capabilities.

To fulfill this case, in the future, we can have two APIs, `ApplicationTrafficPolicy` and `TrafficPolicy`.
I saw the similar case in the TLSBackendPolicy https://gateway-api.sigs.k8s.io/geps/gep-1897/#api

For today, we can just simplify and have one.